File: src/main/java/com/hoolix/processor/models/FilebeatEvent.java
package com.hoolix.processor.models;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.elasticsearch.action.index.IndexRequest;
import org.joda.time.DateTimeZone;
import org.joda.time.chrono.ISOChronology;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.io.IOException;
import java.io.Serializable;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Hoolix 2017
 * Created by simon on 1/1/17.
 */
@AllArgsConstructor
@Getter
public class FilebeatEvent implements Serializable, Event {
    private static final long serialVersionUID = -1061534942493817146L;
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final DateTimeFormatter dateTimeFormatter = DateTimeFormat
            .forPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
            .withLocale(Locale.ENGLISH)
            .withZone(DateTimeZone.UTC)
            .withChronology(ISOChronology.getInstanceUTC());

    @JsonProperty("@timestamp")
    private final String timestamp;

    private final Beat beat;

    private final Map<String, String> fields;

    @JsonProperty("input_type")
    private final String inputType;

    private final String message;

    private final long offset;

    private final String source;

    private final List<String> tags;

    private final String type;

    public static FilebeatEvent fromJsonString(String json) throws IOException {
        return objectMapper.readValue(json, FilebeatEvent.class);
    }

    @Override
    public String getIndexName() {
        return fields.getOrDefault("token", "_default_") + "." + getType();
    }
    @Override
    public IndexRequest toIndexRequest() {
        return new IndexRequest(getIndexName(), getType());
    }

    @Override
    public Map<String, Object> toPayload() {
        Map<String, Object> payload = new HashMap<>();
        payload.put("@timestamp", dateTimeFormatter.parseMillis(timestamp));
        payload.put("beat", beat);
        payload.put("fields", fields);
        payload.put("message", message);
        payload.put("offset", offset);
        payload.put("source", source);
        payload.put("tags", tags);
        payload.put("type", type);
        return payload;
    }
}

File: src/main/scala/com/hoolix/processor/models/FileBeatEvent.scala
package com.hoolix.processor.models

//import spray.json._
//import DefaultJsonProtocol._
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.joda.time.DateTimeZone
import org.joda.time.chrono.ISOChronology
import org.joda.time.format.DateTimeFormat
import org.joda.time.format.DateTimeFormatter
import java.util.Locale

/**
  * Created by peiyuchao on 2017/1/5.
  */
case class FileBeatEvent(
  timestamp: String,
  beat: Beat,
  fields: Map[String, String],
  inputType: String,
  message: String,
  offset: Long,
  source: String,
  tags: Seq[String],
  _type: String
) extends Event {

  def toPayload: collection.mutable.Map[String, Any] = {
    val payload = collection.mutable.Map(
      "timestamp" -> FileBeatEvent.dateTimeFormatter.parseMillis(timestamp),
      "beat" -> beat,
      "fields" -> fields,
      "inputType" -> inputType,
      "message" -> message,
      "offset" -> offset,
      "source" -> source,
      "tags" -> tags,
      "type" -> _type
    )
    payload
  }

  override def indexName = s"${_type}" //add more stuff
  override def indexType = _type
  override def docId = ???
  override def toIndexRequest = ???
}

object FileBeatEvent {
  val dateTimeFormatter: DateTimeFormatter = DateTimeFormat
    .forPattern ("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    .withLocale(Locale.ENGLISH)
    .withZone(DateTimeZone.UTC)
    .withChronology(ISOChronology.getInstanceUTC)

  def fromJsonString(json: String): FileBeatEvent = {

//    println(json)
    implicit val formats = DefaultFormats

//    println(parse(json).transformField({
//      case ("@timestamp", x) => ("timestamp", x)
//      case ("input_type", x) => ("inputType", x)
//      case ("type", x) => ("_type", x)
//    }))

//    println(parse(json).transformField({
//      case ("@timestamp", x) => ("timestamp", x)
//      case ("input_type", x) => ("inputType", x)
//      case ("type", x) => ("_type", x)
//    }).extract[FileBeatEvent])
//    parse(json).extract[FileBeatEvent]

    val fileBeatEvent = parse(json).transformField({
      case ("@timestamp", x) => ("timestamp", x)
      case ("input_type", x) => ("inputType", x)
      case ("type", x) => ("_type", x)
    }).extract[FileBeatEvent]
    fileBeatEvent


    //    import FileBeatEventJsonProtocol._
    //    json.parseJson.convertTo[FileBeatEvent]
  }

}

//object FileBeatEventJsonProtocol extends DefaultJsonProtocol {
////  jsonFormat()
//  implicit val fileBeatEventFormat = jsonFormat9(FileBeatEvent.apply)
////  implicit val userDbEntryFormat = Json.format[UserDbEntry]
////    implicit val fileBeatEventFormat = jsonFormat9("timestamp", "beat", "fields", "input_type", "message", "offset", "source", "tags", "type")
////    implicit val fileBeatEventFormat = jsonFormat(FileBeatEvent)
////    implicit val format = JsonFormat[FileBeatEvent]
//
//}